### DISCUSSION TITLE
Error while trying to run a GNN on a custom Heterogenous Graph

### DISCUSSION THREAD

**Discussion by DominikPKaiser** [pyg-team/pytorch_geometric#3569] [Q&A] [answered]
Hi all,

I'm quite new to the topic of graph neural networks and face an error when trying to run a GNN on a Heterogenous Graph for label classification.

I tried to redo this example on a custom Dataset.
https://pytorch-geometric.readthedocs.io/en/latest/notes/heterogeneous.html

My data looks like follows:

```
HeteroData(
  node_1={
    x=[678, 6], #Node with features 
    y=[678], #Node label i want to predict
    train_mask=[678],
    test_mask=[678]
  },
  node_2={ x=[326, 3] }, #Node with features 
  node_3={ x=[555, 2] }, #Node with features 
  (node_1, rel_1, node_2)={ edge_index=[2, 326] }, #Edges
  (node_3, rel_2, node_2)={ edge_index=[2, 814] }, #Edges
  (node_2, rev_rel_1, node_1)={ edge_index=[2, 326] }, #Edges
  (node_2, rev_rel_2, node_3)={ edge_index=[2, 814] } #Edges
)
```
So I tried to copy the data format from the example, which data structure looks like follows:
```
HeteroData(
  paper={
    x=[736389, 128],
    year=[736389],
    y=[736389],
    train_mask=[736389],
    val_mask=[736389],
    test_mask=[736389]
  },
  author={ x=[1134649, 128] },
  institution={ x=[8740, 128] },
  field_of_study={ x=[59965, 128] },
  (author, affiliated_with, institution)={ edge_index=[2, 1043998] },
  (author, writes, paper)={ edge_index=[2, 7145660] },
  (paper, cites, paper)={ edge_index=[2, 10792672] },
  (paper, has_topic, field_of_study)={ edge_index=[2, 7505078] },
  (institution, rev_affiliated_with, author)={ edge_index=[2, 1043998] },
  (paper, rev_writes, author)={ edge_index=[2, 7145660] },
  (field_of_study, rev_has_topic, paper)={ edge_index=[2, 7505078] }
)
```

The example works fine and also let itself train. I think my structure is the same.

But on my custom dataset I get the following error when I try to train it:
```
RuntimeError                              Traceback (most recent call last)

<ipython-input-63-62b2e85c9fe9> in <module>()
      1 for epoch in range(1, 3):
----> 2   train()

1 frames

<ipython-input-62-3cbb29ef8a78> in train()
      2     model.train()
      3     optimizer.zero_grad()
----> 4     out = model(data.x_dict, data.edge_index_dict)
      5     mask = data['node_1'].train_mask
      6     loss = F.cross_entropy(out['node_1'][mask], data['node_1'].y[mask])

/usr/local/lib/python3.7/dist-packages/torch/fx/graph_module.py in wrapped_call(self, *args, **kwargs)
    614                     print(generate_error_message(topmost_framesummary),
    615                           file=sys.stderr)
--> 616                 raise e.with_traceback(None)
    617 
    618         cls.__call__ = wrapped_call

RuntimeError: expected scalar type Long but found Float
```

I also checked all my tensors if they are not float (torch.LongTensor or torch.BoolTensor for the mask).
The format of "data.x_dict" and "data.edge_index_dict" from my data looks the same like the given example.

The model I used is the GAT() from the example link, I also tried different models and faced the same error.

Does someone knows a solution for this error? Is something wrong within my data structure?

Thanks for any help!

**Comment by rusty1s** [comment, marked as answer]
This is actually quite hard to tell since `torch.fx` does not really reveal much information about the reasons behind this error. Do you can upload your data somewhere so I can take a look?

**Reply by DominikPKaiser** [reply]
Yes, of course. Do you have access to this file?
https://github.com/DominikPKaiser/VertProj2/blob/main/data.pt

**Reply by rusty1s** [reply]
Thanks. I can reproduce the issue and found the reason of it: `x` is of dtype `int64` while it should be a floating-point tensor:
```python
data['node_1'].x = data['node_1'].x.float()
data['node_2'].x = data['node_1'].x.float()
data['node_3'].x = data['node_1'].x.float()
```
fixes the issue.

**Reply by DominikPKaiser** [reply]
Perfect, that fixed the error and my model runs. Lot of thanks for your work and fast responses!
